File: algorithms/medium/maximum_level_sum_of_a_binary_tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::collections::LinkedList;
use std::rc::Rc;
impl Solution {
    pub fn max_level_sum(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let mut q = LinkedList::new();
        q.push_back(root.unwrap());
        let mut ans = 1;
        let mut level = 1;
        let mut maxSum = std::i32::MIN;
        while q.len() > 0 {
            let mut sum = 0;
            let mut p = LinkedList::new();
            while q.len() > 0 {
                let t = q.pop_front().unwrap();
                let n = t.borrow();
                sum += n.val;
                if let Some(left) = n.left.clone() {
                    p.push_back(left);
                }
                if let Some(right) = n.right.clone() {
                    p.push_back(right);
                }
            }
            q = p;
            if maxSum < sum {
                maxSum = sum;
                ans = level;
            }
            level += 1;
        }
        ans
    }
}

File: algorithms/medium/maximum_level_sum_of_a_binary_tree.kt
/**
 * Example: var ti = TreeNode(5) var v = ti.`val` Definition for a binary tree node. class
 * TreeNode(var `val`: Int) { var left: TreeNode? = null var right: TreeNode? = null }
 */
class Solution {
  fun maxLevelSum(root: TreeNode?): Int {
    val q: ArrayDeque<TreeNode> = ArrayDeque()
    var ans = 1
    var level = 1
    var max_sum = root!!.`val`

    q.add(root!!)

    while (q.size > 0) {
      val size = q.size
      var sum = 0
      for (i in 1..size) {
        val node = q.removeFirst()
        sum += node.`val`
        node.left?.let { q.add(node.left) }
        node.right?.let { q.add(node.right) }
      }

      if (max_sum < sum) {
        max_sum = sum
        ans = level
      }
      level += 1
    }

    return ans
  }
}
